### ISSUE TITLE
"TypeError: no implicit conversion of nil into String" for write_request when user needs to establish/re-establish a valid run_loop is not user-friendly

### ISSUE THREAD

**Issue by TeresaP** [calabash/run_loop#102]
On version 12.0 for client and server

I noticed the following:

1. The user is using the console and the run_loop gets messed up (to where the user needs to do another start_test_server_in_background)
2. User performs a gesture which results in `RunLoop::Core.write_request(run_loop, cmd, logger=nil)` being called
3. `lib/run_loop/core.rb:L498` => `index = run_loop[:index]`; `index` is nil because run_loop does not contain a key for :index.
4. index is referenced later, though it is nil, resulting in:

```
2015-01-08 10:39:39 -0800: Trying write of command :uia.flickOffset('{:x 394, :y 542}', '{:x 2, :y 542}') at index 
TypeError: no implicit conversion of nil into String
	from /Users/foo/.rvm/gems/ruby-2.0.0-p247/gems/run_loop-1.2.0/lib/run_loop/core.rb:505:in `initialize'
	from /Users/foo/.rvm/gems/ruby-2.0.0-p247/gems/run_loop-1.2.0/lib/run_loop/core.rb:505:in `open'
	from /Users/foo/.rvm/gems/ruby-2.0.0-p247/gems/run_loop-1.2.0/lib/run_loop/core.rb:505:in `block in write_request'
	from /Users/foo/.rvm/gems/ruby-2.0.0-p247/gems/run_loop-1.2.0/lib/run_loop/core.rb:503:in `times'
	from /Users/foo/.rvm/gems/ruby-2.0.0-p247/gems/run_loop-1.2.0/lib/run_loop/core.rb:503:in `write_request'
	from /Users/foo/.rvm/gems/ruby-2.0.0-p247/gems/run_loop-1.2.0/lib/run_loop/core.rb:824:in `send_command'
	from /Users/foo/.rvm/gems/ruby-2.0.0-p247/gems/calabash-cucumber-0.12.0/lib/calabash-cucumber/uia.rb:41:in `uia'
	from /Users/foo/.rvm/gems/ruby-2.0.0-p247/gems/calabash-cucumber-0.12.0/lib/calabash-cucumber/uia.rb:430:in `uia_handle_command'
	from /Users/foo/.rvm/gems/ruby-2.0.0-p247/gems/calabash-cucumber-0.12.0/lib/calabash-cucumber/uia.rb:207:in `uia_flick_offset'
	from (irb):17
	from /Users/foo/.rvm/rubies/ruby-2.0.0-p247/bin/irb:13:in `<main>' 
```

I would expect a message to the user about needing to do a start_test_server_in_background or something to that effect.

**Comment by jmoody** [comment]
@TeresaP Snap.  I will take a look.

**Comment by jmoody** [comment]
@TeresaP 

I want to make sure I understand what is happening and what you expect.

> The user is using the console and the run_loop gets messed up (to where the user needs to do another start_test_server_in_background)

Can you describe this bad state?  Is it that the irb stack has a error?  Some query/gesture failed to work as expected?  Or is it that the run loop is _in_ a bad state; you, as the user, have no knowledge that something is wrong until Step 2?

> I would expect a message to the user about needing to do a start_test_server_in_background or something to that effect.

Are you suggesting that _if_ index is nil, an informative error should be raised rather than this generic "no implicit conversion of nil into String" exception?

My gut is telling me that `run_loop[:index] => nil` is the result of a bug.

@TeresaP Are you calling `> console_attach` at any point?

@krukow Can you shed some light on this?

**Comment by jmoody** [comment]
@TeresaP Ha! This is a duplicate:

**When :uia_strategy => :host, `console_attach` attaches, but throws exception when performing gestures.** [#638](https://github.com/calabash/calabash-ios/issues/638)

And it is a bug.

You don't need to answer my questions above.

**Comment by jmoody** [comment]
The `index` can be set in `console_attach`; I think it can be set to 0 without any trouble.

However, the `repl_path` is also nil after `console_attach`.

```
 def self.write_request(run_loop, cmd, logger=nil)
      repl_path = run_loop[:repl_path]
      index = run_loop[:index]
      cmd_str = "#{index}:#{escape_host_command(cmd)}"
      should_log = (ENV['DEBUG'] == '1')
```

**Comment by TeresaP** [comment]
Thanks @jmoody  :)

**Comment by jmoody** [comment]
@TeresaP @krukow 

In order for the :host strategy to use `console_attach`, we need to be able to retrieve the current run-loop hash.

```
{
 :pid=>58290,
 :index=>3,
 :uia_strategy=>:host,
 :udid=>"193688959205dc7eb48d603c558ede919ad8dd0d",
 :app=>"com.xamarin.chou-cal",
 :repl_path=>
  "/var/folders/bx/5wplnzxx7492fynry0m_hw080000gp/T/run_loop20150109-58079-1952w1i/repl-cmd.pipe",
 :log_file=>
  "/var/folders/bx/5wplnzxx7492fynry0m_hw080000gp/T/run_loop20150109-58079-1952w1i/run_loop.out",
 :results_dir=>
  "/var/folders/bx/5wplnzxx7492fynry0m_hw080000gp/T/run_loop20150109-58079-1952w1i",
 :initial_offset=>99697
}
```

The key pieces of information are:

* index
* initial_offset

We can deduce everything else by inspecting the process information (using `ps`).

The only solution, I think, is to save the run_loop hash to disk and read it back into memory in Calabash::Cucumber::Launcher.attach.

@krukow This might be an interesting time to explore this.  We have been discussing the idea of caching information between runs to improve performance.  We might use this example as way to test the waters.

**Comment by jmoody** [comment]
Will be available in run-loop 1.2.3 and calabash-ios 0.12.1.

**Comment by jmoody** [comment]
Fixed in calabash-ios 0.12.2 and run-loop 1.2.6.
